Fixing Attach Volume to Instance

rgildein opened the pull request:
This tutorial refers to private playbook used by BootStack (NovaComputeFixVolumeAttach), including prerequisites.

Commit:
Fixing Attach Volume to Instance

This tutorial refers to [1], including prerequisites.

[1]: https://wiki.canonical.com/CDO/IS/Bootstack/Playbooks/NovaComputeFixVolumeAttach

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
```diff
@@ -0,0 +1,336 @@
+
+### OpenStack
+
+Deploy the openstack-base bundle with the custom overlay.
```
Link to https://jaas.ai/openstack-base

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
```diff
@@ -0,0 +1,336 @@
+[comment]: <> (How To Perform Fixing Attach Volume to Instance)
```
Suggestion: How To Fix A Volume Attachment Failure Against An OpenStack Instance

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
```diff
@@ -0,0 +1,336 @@
+[comment]: <> (How To Perform Fixing Attach Volume to Instance)
+
+## pre-install
```
Requirements? Have a look to https://github.com/canonical/doc-docs/pull/3/files for similar terminology (like the "Requirements" section).

rgildein replied:
Thanks, I divided it into "Introduction" & "Requirements".

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
```diff
@@ -0,0 +1,336 @@
+
+Duration: 120:00
+
+This description of how the deployment of `charmed kubernetes` is performed on a OpenStack server, where OpenStack is deployed to another OpenStack. (openstack-on-openstack).
```
Suggestion: This tutorial has been tested on a CDK (link to the CDK bundle) environment deployed on an OpenStack (link to openstack base) environment. The OpenStack environment was deployed on top of another managed OpenStack service.

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
````diff
@@ -0,0 +1,336 @@
+juju deploy ./openstack-bundles/stable/openstack-base/bundle.yaml --overlay ./openstack-bundles/stable/overlays/openstack-base-virt-overlay.yaml --overlay ./memory-overlay.yaml
+```
+
+##### initialize and unseal vault and create self-signed root CA
````
* Start with upper case (Initialize)
* "create a self-signed root CA"

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
````diff
@@ -0,0 +1,336 @@
+3. Initialize the vault deployment.
+   ```bash
+   export VAULT_ADDR="http://$(juju status | grep vault/0 | tr -s " " | cut -d" " -f5):8200"
+   vault operator init -key-shares=5 -key-threshold=3 > vault-keys
````
Would it be possible to remove the "1." item and use `juju ssh vault/0` and then `export VAULT_ADDR="http://127.0.0.1:8200"` instead? The vault unit will already have the vault snap installed so no need of manually installing the snap anywhere else.

rgildein replied:
You're right, I changed it to run on the `vault/0` unit.

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
````diff
@@ -0,0 +1,336 @@
+   juju run-action --wait vault/leader generate-root-ca
+   ```
+   
+##### add external L3 connectivity
````
This section needs an explanation on why the following steps are needed. Why 3 ports need to be manually created? Is that an issue related to Serverstack or to any "OpenStack managed service running underneath our testing env (CDK over OpenStack over a managed OpenStack)"

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
````diff
@@ -0,0 +1,336 @@
+   juju config ovn-chassis bridge-interface-mappings="$BRIDGE_INTERFACE_MAPPING"
+   ```
+   
+##### final configuration
````
"Final" (start with upper case)

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
```diff
@@ -0,0 +1,336 @@
+The final step is to configure juju to use deployed OpenStack as cloud and configure network on OpenStack.
+
+[note status="OpenStack credentials"]
+Don't forget to get credentials for the new OpenStack. (e.g. `source openrcv3_project`)
```
Where does `openrcv3_project` come from? A user following this tutorial should be able to follow all the steps without any further context (or without searching for other docs... except if they are linked in this tutorial).

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
```diff
@@ -0,0 +1,336 @@
+Don't forget to get credentials for the new OpenStack. (e.g. `source openrcv3_project`)
+[/note]
+
+1. To access the OpenStack cloud is necessary to set up flavors, images, networks and security. Detailed instructions can be found on [openstack-base][OpenStackBase] in the "Accessing the cloud" section.
```
"In order to use the managed OpenStack service, several configurations are required: flavors, images, networks and security groups. Detailed ..."

auria replied:
Going through the tutorial, trying to reproduce it myself, and seeing the following section with the network details in yaml format, I wonder how I could translate that into applicable configurations in the environment. I wonder if it would be best to share the needed commands. For example (taken from the "Accessing the cloud" section of the openstack-base README instead of making the tutorial reader jump here and there):

```
source ~/openstack-bundles/stable/openstack-base/openrc
openstack catalog list
openstack image list
curl https://cloud-images.ubuntu.com/focal/current/focal-server-cloudimg-amd64.img |     openstack image create --public --container-format=bare --disk-format=qcow2 focal
openstack network create --external --provider-network-type flat --provider-physical-network physnet1 ext_net
openstack subnet create --subnet-range 10.5.0.0/16 --no-dhcp --gateway 10.5.0.1 --network ext_net --allocation-pool start=10.5.0.200,end=10.5.0.254 ext
openstack network create internal
openstack subnet create --network internal --subnet-range 192.168.21.0/24 --dns-nameserver 10.245.168.6 internal_subnet
openstack router create provider-router
openstack router set --external-gateway ext_net provider-router
openstack router add subnet provider-router internal_subnet
openstack flavor create --ram 2048 --disk 0 --ephemeral 0 m1.small
openstack keypair create --public-key ~/.ssh/id_rsa.pub mykey
openstack server create --image focal --flavor m1.small --key-name mykey --network internal focal-test
# Secgroup rules for the admin project default secgroup
openstack security group list
openstack project list
openstack security group rule list a3df0159-c7b4-4542-a30d-cc0781c3c44a
openstack security group rule create --proto icmp --remote-ip 0/0 a3df0159-c7b4-4542-a30d-cc0781c3c44a
openstack security group rule create --proto tcp --remote-ip 0/0 --dport 22 a3df0159-c7b4-4542-a30d-cc0781c3c44a
openstack security group rule create --proto tcp --remote-ip 0/0 --dst-port 22 a3df0159-c7b4-4542-a30d-cc0781c3c44a
```

I was not able to ssh into the private IP, by routing via the Virtual router. I had to create a floating IP:

```
FIP=$(openstack floating ip create -f value -c floating_ip_address ext_net)
openstack server add floating ip focal-test $FIP
```

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
```diff
@@ -0,0 +1,336 @@
+
+2. The last thing is to add a new OpenStack cloud to Juju, create new credentials and create a controller. See the [OpenStack-cloud][OpenStackCloud] page for more information.
+    [note status="OpenStack metadata"]
+    It's necessary to create metadata for initializing a Juju cloud environment (creating the controller).
```
I wonder if it may be worth adding a link to the official documentation about Juju metadata?

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
```diff
@@ -0,0 +1,336 @@
+
+### Kubernetes
+
+Deployed charmed kubernetes on top of OpenStack with `charm-openstack-integrator` act as proxy for OpenStack and provide a control interface.
```
"Deploy CDK (link to CDK) on top of OpenStack using charm-openstack-integrator. The integrator will act as a proxy between Kubernetes and OpenStack, and it will provide a control interface (eg. creating persistent volumes)."

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
````diff
@@ -0,0 +1,336 @@
+juju deploy charmed-kubernetes --overlay openstack-overlay.yaml --trust
+```
+
+Before configuration `openstack-integrator` it is necessary to source into the Openstack credential file. e.g. `source novarc`
````
"Before configuring ..., load the OpenStack credentials as environment variables e.g. `source novarc`"

However, it needs to be clear at the beginning of the tutorial where the `novarc` file, as well as the previously mentioned `openrcv3_project` file can be found.

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
```diff
@@ -0,0 +1,336 @@
+
+##### install
+
+The most common way to install `kubectl` in ubuntu is `snap`.
```
"is as a snap."

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
````diff
@@ -0,0 +1,336 @@
+[note status="Proxy"]
+A proxy must be set on OpenStack.
+```bash
+juju config containerd http_proxy=http://squid.internal:3128 https_proxy=http://squid.internal:3128
````
It's a good idea to mention the possible need of proxy settings, but not every user will need one. And certainly, not the one specified here. "squid.internal" should be replaced by `<proxy-ip>`

rgildein replied:
You are absolutely right, this is sensitive internal data.

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
```diff
@@ -0,0 +1,336 @@
+This is due to [LP#1847361][#1847361]. In summary, this bug causes the new **QEMU** version to not recognize the old **QEMU** modules for attaching volumes to VMs. 
+This bug is especially bad in Kubernetes deployments on top of Openstack. A pod will get stuck attempting to create container when it tries to attach the PVC (and by consequence, attach cinder volume to k8s worker VM). This playbook intends to explain how to identify if the **QEMU** bug is the root problem, how to fix, and proper procedure for handling the k8s situation.
+
+### How To Identify
```
I know this section name comes from the playbook. A better section title could be "Identify attachment failures"

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
```diff
@@ -0,0 +1,336 @@
+
+### K8s Attachment Issues
+
+Extra steps must be taken when this problem manifests in a **K8s** cloud. Care must be taken to drain the **K8s** worker before actioning the instance. It may be possible to migrate the **K8s** worker instance without draining, but this hasn't been explicitly tested. These steps are taken from the [K8s Worker Maintenance Playbook][K8sWorkerMaintenance].
```
Once the other tutorial is published, we will link it here and remove the rest, shared below.

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
```diff
@@ -0,0 +1,336 @@
+
+### Details
+
+[CategoryISTeam][CategoryISTeam]
```
This link should be removed. It only makes sense in the wiki.

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
```diff
@@ -0,0 +1,336 @@
+...
+
+[#1847361]: https://bugs.launchpad.net/ubuntu/+source/qemu/+bug/1847361
+[K8sWorkerMaintenance]: https://wiki.canonical.com/CDO/IS/Bootstack/Playbooks/K8sWorkerMaintenance
```
This link will be against juju.is/tutorials/how-to-perform-k8s-maint... or whichever URL is created when https://discourse.charmhub.io/t/how-to-perform-maintenance-on-a-kubernetes-worker/3910 is published in juju.is/tutorials

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
```diff
@@ -0,0 +1,336 @@
+
+[#1847361]: https://bugs.launchpad.net/ubuntu/+source/qemu/+bug/1847361
+[K8sWorkerMaintenance]: https://wiki.canonical.com/CDO/IS/Bootstack/Playbooks/K8sWorkerMaintenance
+[CategoryISTeam]: https://wiki.canonical.com/CategoryISTeam
```
Remove this

auria reviewed (changes requested):
I've added several suggestions. I see there are 2 main sections, which would be listed on the left menu like at https://juju.is/tutorials/get-started-charmed-kubernetes. We may want to consider more than 2 sections, or make subsections as principal sections.

Also, I wanted to share the consideration of making the tutorial self-sufficient (no need to check other docs or tutorials except if linked in this tutorial) and easy to follow for someone without any further context (eg. internet user, not a canonical employee, and no previous knowledge of openstack-charmers repos, bundles or specific test environments).

Overall, the content heads on the right track. Thank you!

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
````diff
@@ -0,0 +1,336 @@
+   ```
+4. Unseal all keys and generate root CA.
+   ```bash
+   for key in $(cat vault-keys | grep -E "Unseal Key .*: " | cut -c15-)
````
I think `| head -3` can be added at the end, since 3 keys are enough to unseal vault.

OTOH, the "grep" command can get a file to read, so "cat" wouldn't be needed.

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
````diff
@@ -0,0 +1,336 @@
+   ```
+5. Authorize the vault unit.
+   ```bash
+   export VAULT_TOKEN=$(cat vault-keys | grep -E "Initial Root Token: " | cut -c21-)
````
Same about "grep" reading from "vault-keys" instead of from stdin (so the "cat" command can be skipped).

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
````diff
@@ -0,0 +1,336 @@
+Deployed charmed kubernetes on top of OpenStack with `charm-openstack-integrator` act as proxy for OpenStack and provide a control interface.
+
+```bash
+juju deploy charmed-kubernetes --overlay openstack-overlay.yaml --trust
````
"juju bootstrap" step is missing. In my case, I used:

```
# Allow multiple units to land in 3 nova-compute units
juju config nova-cloud-controller disk-allocation-ratio=10
# Juju controller needs mem=3584M by default
openstack flavor create --ram 3584 --vcpus 2 --disk 20 --ephemeral 0 m1.small
juju metadata generate-image -d ./simplestreams -i $IMG_ID -s focal -r $OS_REGION_NAME -u $OS_AUTH_URL
juju bootstrap --metadata-source /home/ubuntu/testing/simplestreams  --config external-network=$EXT_NET --config network=$INTERNAL --config use-floating-ip=true --config use-default-secgroup=true --config default-series=focal cloud-name cloud-name

# flavor to be used by other models
openstack flavor create --ram 2048 --vcpus 1 --disk 10 m1.micro
```
Note: I had to copy the `$OS_CACERT` file to `/usr/local/share/ca-certificates/cloud-name.crt` and run `sudo update-ca-certificates` or Juju would not be able to verify the $OS_AUTH_URL endpoint from the overcloud.

OTOH, the `$OS_CACERT` and rest of `OS_` envvars were used to run `juju add-cloud`. A config example can be found at:

```
$ juju show-cloud os-aluria

Client cloud "os-aluria":

defined: local
type: openstack
description: Openstack Cloud
auth-types: [userpass]
endpoint: https://10.5.0.8:5000/v3
credential-count: 1
regions:
  RegionOne: {}
ca-credentials:
- |
  -----BEGIN CERTIFICATE-----
  MIIDazCCAlOgAwIBAgIUe1/iYf6EtsDUguj25agMNtpglwwwDQYJKoZIhvcNAQEL
  BQAwPTE7MDkGA1UEAxMyVmF1bHQgUm9vdCBDZXJ0aWZpY2F0ZSBBdXRob3JpdHkg
  KGNoYXJtLXBraS1sb2NhbCkwHhcNMjAxMjE1MDcyNjIwWhcNMzAxMjEzMDYyNjQ5
  WjA9MTswOQYDVQQDEzJWYXVsdCBSb290IENlcnRpZmljYXRlIEF1dGhvcml0eSAo
  Y2hhcm0tcGtpLWxvY2FsKTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEB
  AL5ssaJ/cxitgU5s8o+8Ee+7wtH3DoXMEBAAsh2dQehy8Nms1l0Ha/fuT7vTMFNh
  Y+8O0BFd0Md+FQ2ODqzBXz5jxe+gusr798YiuUnQGouim9x1E17W/odaqGMCxekx
  iNBfC7mMkTZaSvEx9zxkITVO94H+i6JrXvdwGyEmR3JjtdcsqVcKNf9kBrH+buX4
  c55etPW4WZOob2Bv/n7SzHfdJSMJzuKMUR0bm927zozZOwHNGw5lGz5SgYUkHB2i
  yo+YRg5y3t2uHGC4nIPnw60WndxJcLpSm0VCCNigSEeZlQipbgAp1x6uxjNxrA5H
  9hqTc3pMpn777i8nra+OWE8CAwEAAaNjMGEwDgYDVR0PAQH/BAQDAgEGMA8GA1Ud
  EwEB/wQFMAMBAf8wHQYDVR0OBBYEFGy/EdwM9hh4Va6VbxHX8frUQic1MB8GA1Ud
  IwQYMBaAFGy/EdwM9hh4Va6VbxHX8frUQic1MA0GCSqGSIb3DQEBCwUAA4IBAQB6
  skBs1G5tutAER+odFcGVzip/1JHMjJKQOtjirJRrVn1lH0bh9FgA4GpgsX9wedJc
  cLwZOSIvblYq1NjqqUZFf7p29knix8K0JPznwh9SEhvOkB9TKHN2YWHrTxpfni3S
  3mbNehUaRrzqj+nTYdaKWKphUnjE9ZrHHWsftM1KS/Rq5uiMA/B/UuSYy5FYv25q
  LOB6Twvi/wgOkMsFx9ufW8/VEprD1o2dPHPxns/nkHwc/LsCx6Ogioi1jSGvoQHV
  o8JPwZo/IPZ5itMRmDhNZU4dbWBMIIfo93nL8oI12AqJCM8yLNGQnNEELwmg+Cn3
  9an6HY28N9Ishpopyz9S
  -----END CERTIFICATE-----
```

I also ran `juju add-credential os-aluria -f credentials.yaml` using the following `credentials.yaml` file:

```
credentials:
  os-aluria:
    admin:
      auth-type: userpass
      password: REDACTED
      tenant-name: admin
      username: admin
      project-domain-name: admin_domain
      user-domain-name: admin_domain
```

auria reviewed (changes requested):
I've added some feedback and a few examples that I have been able to make them work in serverstack.

The DNS used in the tenant (internal) network is the same as the instances in your serverstack tenant use (`10.5.0.2`). The other IP used and documented here returns `refused`.

Since the openstack-on-serverstack environment runs vault, a OS_CACERT needs to get configured on the Juju client or it won't be able to create the controller. I've also added an example of `clouds.yaml` configuration to bootstrap the controller.

Two flavors need to be used:
* m1.micro, 2GB ram, 1 vCPU and 10G disk
* m1.small 3GB ram, 2 vCPU and 20G disk for the controller

The controller needs to be bootstrapped with several extra configurations, like the use of the default security group and the use of a floating IP or we won't be able to ssh into it.

For other models, units can be ssh'ed like `juju ssh --proxy <unitid>`.

I have been able to spawn 9 machines. This is by using 3 nova-compute units of 8GB RAM, and changing the `disk-allocation-ratio` and `ram-allocation-ratio`, as well as increasing the default OpenStack quotas:
```
openstack quota set --instances 20 --secgroups 40 <admin-project-id>
openstack quota show <admin-project-id>
```

As a reminder, if an instance on top of the openstack-on-serverstack environment uses a floating IP, it will be reachable via ssh. Otherwise, `juju ssh --proxy` will be needed.

auria commented:
BTW, I also think this tutorial should be split in 3:
 * How to deploy openstack-on-openstack, and bootstrap a Juju env on top of it.
 * How to fix the qemu issue, the scope of this PR
 * How to manage Kubernetes to make safe changes on qemu (helping yourself with [Martin's tutorial](https://discourse.charmhub.io/t/how-to-perform-maintenance-on-a-kubernetes-worker/3910)

By doing that, people would be able to skip some parts (ie. if they already have a Kubernetes environment running on their own OpenStack environment).

Commit:
Applied changes suggested by Alvaro

Wrote more details about the deployment of OpenStack on OpenStack and
divided it into two documents.

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
```diff
@@ -0,0 +1,371 @@
+[comment]: <> (How to deploy openstack-on-openstack, and bootstrap a Juju env on top of it)
```
Correct the .md file name: "openstakc" to "openstack"

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
```diff
@@ -0,0 +1,371 @@
+
+### OpenStack
+
+Deploy the [OpenStack-base][openstack-base] bundle with the custom overlay.
```
I think it would be beneficial a quick clarification on why the "storage" definition is needed. The latest version of Ceph needs disks partitions to create OSDs. Filestore is not supported, so OSDs can't be run on raw files.

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
````diff
@@ -0,0 +1,371 @@
+juju deploy ./openstack-bundles/stable/openstack-base/bundle.yaml --overlay ./openstack-bundles/stable/overlays/openstack-base-virt-overlay.yaml --overlay ./resource-overlay.yaml
+```
+
+##### Initialize and unseal the Vault and create a self-signed root CA
````
Suggestion: "unseal Vault" or "unseal the Vault units"

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
````diff
@@ -0,0 +1,371 @@
+   ```bash
+   juju ssh vault/0
+   ```
+2. Initialize the vault deployment.
````
"Initialize Vault" (more info: https://www.vaultproject.io/docs/commands/operator/init)

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
````diff
@@ -0,0 +1,371 @@
+   ```
+4. Unseal all keys and generate root CA.
+   ```bash
+   for key in $(grep -E "Unseal Key .*: " vault-keys | cut -c15-)
````
I think `| tail -3` could also be used, since the minimum number of keys to unseal Vault is 3.

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
````diff
@@ -0,0 +1,371 @@
+       vault operator unseal $key
+   done
+   ```
+5. Initial TTL to associate with the token.
````
Since "TTL" is also the arg used in the command below, I would suggest explaining what this step does. "Create a temporary token which will expire after 10 minutes" (more info: https://www.vaultproject.io/docs/commands/token/create)

auria replied:
this woudl also be step 4, not 5.

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
````diff
@@ -0,0 +1,371 @@
+   export VAULT_TOKEN=$(juju ssh vault/0 'grep -E "Initial Root Token: " vault-keys | cut -c21-')
+   juju run-action --wait vault/leader authorize-charm token=$VAULT_TOKEN
+   juju run-action --wait vault/leader generate-root-ca
+   ```
````
I think there is a step missing in this process. At this point, `vault/0` Vault service has been unsealed. However, the other 2 units also run a Vault service which needs to be unsealed with the same keys saved on step 2.

Once all Vault services in the cluster have been unsealed, step 7 (authorize the vault unit) could be run.

rgildein replied:
There is only one vault unit in the openstack-base package. Or am I missing something?

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
````diff
@@ -0,0 +1,371 @@
+   vault token create -ttl=10m
+   ```
+6. Detach from the `vault/0` unit. 
+5. Authorize the vault unit.
````
This is step 6, not 5.

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
````diff
@@ -0,0 +1,371 @@
+   export VAULT_ADDR="http://127.0.0.1:8200"
+   vault operator init -key-shares=3 -key-threshold=3 > vault-keys
+   ```
+4. Unseal all keys and generate root CA.
````
this is step 3, not 4.

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
````diff
@@ -0,0 +1,371 @@
+   export VAULT_TOKEN=$(grep -E "Initial Root Token: " vault-keys | cut -c21-)
+   vault token create -ttl=10m
+   ```
+6. Detach from the `vault/0` unit. 
````
step 5, not 6

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
```diff
@@ -0,0 +1,371 @@
+##### Add external L3 connectivity
+
+This section deals with the prerequisites for network configuration and is required for the full functionality of 
+the deployed OpenStack. It guarantees a connection between individual units.
```
"It guarantees network traffic between individual units" ?

rgildein replied:
"It provides network traffic between individual units." Reading this now, I don't like the word guarantee.

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
````diff
@@ -0,0 +1,371 @@
+   ```bash
+   export ADMIN_NET=$(openstack network list | grep admin_net | cut -d"|" -f3 | tr -d " ")
+   ```
+2. Create new port for each nova-compute unit. (by default there are 3 units)
````
"Create a new port..."

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
```diff
@@ -0,0 +1,371 @@
+   for i in 0 1 2
+   do
+       # create new port
+       export NEW_PORT_$i_ID=$(openstack port create --network $ADMIN_NET data-port | grep ' id ' | cut -d"|" -f3 | tr -d " ")
```
I'd suggest exploring `openstack port create -f shell --network ....` as the output will be formatted in shell format and the same variable can be used (`$PORT_ID` or something like that)

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
```diff
@@ -0,0 +1,371 @@
+       echo "  create new port with id $NEW_PORT_$i_ID"
+    
+       # get ovn chassis ip
+       export OVN_CHASSIS_$i_IP=$(juju status | grep ovn-chassis/$i | tr -s " " | cut -d" " -f5)
```
Suggest to use the json format to parse it in a single pipe. The list of units have a "public-address" key with the value, which I have transformed from a keymap (`{"unit/0": "X.X.X.X", "unit/1": "X.X.X.Y", "unit/2": "X.X.X.Z"}`) to an array (`["X.X.X.X", "X.X.X.Y", "X.X.X.Z"]`) from which we can return each element (0, 1 and 2).
```
$ juju status --format json | jq -r '.applications."ovn-chassis".units | map_values(."public-address") | [.[]] | .[0]'
10.5.0.27
$ juju status --format json | jq -r '.applications."ovn-chassis".units | map_values(."public-address") | [.[]] | .[1]'
10.5.0.16
$ juju status --format json | jq -r '.applications."ovn-chassis".units | map_values(."public-address") | [.[]] | .[2]'
10.5.0.13
```

rgildein replied:
I'm testing it now, if the unit is blocked, it doesn't work.
```bash
➜  NovaComputeFixVolumeAttach juju status --format json | jq -r '.applications."ovn-chassis"'
{
  "charm": "cs:ovn-chassis-6",
  "series": "focal",
  "os": "ubuntu",
  "charm-origin": "jujucharms",
  "charm-name": "ovn-chassis",
  "charm-rev": 6,
  "charm-version": "404a44d",
  "charm-profile": "juju-k8-ovn-chassis-6",
  "exposed": false,
  "application-status": {
    "current": "error",
    "message": "hook failed: \"ovsdb-relation-changed\"",
    "since": "28 Jan 2021 13:20:25+01:00"
  },
  "relations": {
    "certificates": [
      "vault"
    ],
    "nova-compute": [
      "nova-compute"
    ],
    "ovsdb": [
      "ovn-central"
    ]
  },
  "subordinate-to": [
    "nova-compute"
  ],
  "version": "20.03.1",
  "endpoint-bindings": {
    "": "alpha",
    "amqp": "alpha",
    "certificates": "alpha",
    "data": "alpha",
    "juju-info": "alpha",
    "nova-compute": "alpha",
    "nrpe-external-master": "alpha",
    "ovsdb": "alpha",
    "ovsdb-subordinate": "alpha"
  }
}
```

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
```diff
@@ -0,0 +1,371 @@
+       echo "  found ovn-chassis/$i ip address $OVN_CHASSIS_$i_IP"
+    
+       # get ovn chassis id
+       export OVN_CHASSIS_$i_ID=$(openstack server list | grep -E $OVN_CHASSIS_$i_IP | cut -d"|" -f3 | tr -d " ")
```
The first command can use `-f value` to remove the pipes and simply use `awk -v IP=$OVN_CHASSIS_IP '/IP/ {print $3}'` (the third column).

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
```diff
@@ -0,0 +1,371 @@
+       echo "  add port $NEW_PORT_$i_ID to server $OVN_CHASSIS_$i_ID"
+    
+       # get port mac address
+       export NEW_PORT_$i_MAC_ADDRESS=$(openstack port show $NEW_PORT_$i_ID | grep mac_address | cut -d"|" -f3 | tr -d " ")
```
Please explore `openstack port show -fvalue -c mac_address $PORT_ID` as an option to retrieve a single value

auria replied:
However, if `-f shell` was used when the port was created, you may have that value already stored in VAR=VALUE format.

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
```diff
@@ -0,0 +1,371 @@
+   
+##### Final configuration
+
+The final step is to configure Juju to use deployed OpenStack as cloud and configure a OpenStack.
```
"to use the deployed OpenStack environment as cloud provider, and configure it."

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
```diff
@@ -0,0 +1,371 @@
+2. The last thing is to add a new OpenStack cloud to Juju, create new credentials and create a controller. 
+   See the [OpenStack-cloud][openstack-cloud] page for more information.
+   
+   1. Add OpenStack as a cloud to Juju.
```
"Add openstack as a Juju cloud provider"

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
````diff
@@ -0,0 +1,371 @@
+
+      ```
+      [/note]
+   2. Add credentials to OpenStack cloud.
````
"Add credentials to the "openstack" cloud"

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
````diff
@@ -0,0 +1,371 @@
+          user-domain-name: admin_domain
+          version: "3"
+      ```
+      [\note]
````
Should this be "[/note]" ?

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
````diff
@@ -0,0 +1,371 @@
+      ```
+      [note status="Proxy"]
+      If a proxy is required on the servers where you deploy it, also use 
+      `--model-default program http-proxy=<proxy_ip>:<proxy_port> https-proxy=<proxy_ip>:<proxy_port>`. 
````
What is "program"? On the other hand, I don't recall if `--model-default` is needed for each variable or a single call to the argument followed by multiple variables is ok.

rgildein replied:
It's a good question, I don't know where this "program" came from. And no, I used it without recall it and it works well.

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
````diff
@@ -0,0 +1,371 @@
+```
+
+[note status="OpenStack credentials"]
+Before configuration `openstack-integrator` load the OpenStack credentials as environment variables. 
````
Suggestion: "In order to configure the `openstack-integrator` application, the OpenStack credentials need to be loaded as environment variables."

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
```diff
@@ -0,0 +1,371 @@
+
+[note status="OpenStack credentials"]
+Before configuration `openstack-integrator` load the OpenStack credentials as environment variables. 
+(e.g. `source novarc`, look at "Requirements" section)
```
"look at _the_ "Requirements" section"

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
````diff
@@ -0,0 +1,371 @@
+[/note]
+
+```bash
+juju config openstack-integrator auth-url=$OS_AUTH_URL region=$OS_REGION_NAME username=$OS_USERNAME user-domain-name=$OS_USER_DOMAIN_NAME password=$OS_PASSWORD project-name=$OS_PROJECT_NAME  project-domain-name=$OS_PROJECT_DOMAIN_NAME
````
Is this needed if `juju trust` has been run? It seems like `juju deploy` runs with `--trust`, which should give access to the credentials used by the Juju "openstack" cloud to bootstrap the CDK environment.

If `juju deploy --trust` works as expected, the note to source novarc and apply this config would not be needed.

auria replied:
However, I don't think it would be wrong to mention all this as an alternative to "--trust". Maybe a Juju user wants to use different credentials to bootstrap CDK and provide auth access to k8s master and worker units.

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
```diff
@@ -0,0 +1,371 @@
+### Kubectl
+
+
+##### Install
```
This section has 5 "#" and the Kubectl section has 3. Shouldn't the "Install" subsection have 4 "#"?

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
````diff
@@ -0,0 +1,371 @@
+```
+
+[note status="Tunnel"]
+If your OpenStack is not on the local machine, or you are not working directly on the remote machine, you may 
````
"If your OpenStack environment"

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
```diff
@@ -0,0 +1,371 @@
+
+### Cinder volume
+
+##### Create
```
Same: 5 "#" instead of 4?

auria commented on juju-tutorials/how-to-deploy-openstakc-on-openstack.md:
````diff
@@ -0,0 +1,371 @@
+
+```bash
+juju run --application nova-compute 'sudo apt install --reinstall qemu'
+```
````
I think this part about updating QEMU does not belong to this document.

* This document should only describe how to deploy OpenStack on OpenStack.
* A different document should explain how to deploy CDK on OpenStack.
* A third document should explain how to upgrade QEMU, after safely putting k8s workers (running on top of the nova-compute units that we want to upgrade) into maintenance. The part about setting k8s workers in maintenance mode was described in a doc created by Martin.

rgildein replied:
Okay, I'll split these two `.md` files into three,

auria commented on juju-tutorials/how-to-fixing-attach-volume-to-a-k8s-instance.md:
```diff
@@ -0,0 +1,113 @@
+Extra steps must be taken when this problem manifests in a **K8s** cloud. Care must be taken to drain the **K8s** 
+worker before actioning the instance. It may be possible to migrate the **K8s** worker instance without draining, 
+but this hasn't been explicitly tested. These steps are taken 
+from the [K8s Worker Maintenance Playbook][K8sWorkerMaintenance].
```
The "K8s Worker Maintenance Playbook" should be renamed to "K8s Worker Maintenance Tutorial".

OTOH, since that tutorial already explains the steps, I think it should be linked but the steps shouldn't be reproduced in this tutorial. It would be ok to explain _what_ needs to be done (high level explanation) and point to the tutorial that explains _how_ to do it (with all the commands).

auria reviewed (changes requested):
Hi, sorry for the delay.

I have added several suggestions for changes. The tutorials look in a good shape but a little more work needs to be done around grammar corrections, some parts being rephrased and maybe test other oneliners suggested if you think it simplifies data collection (if not, feel free to leave the current oneliners).

Thank you!

Commit:
Divided into 3 tutorials and fix the document style

auria commented on juju-tutorials/how-to-deploy-CDK-on-openstack.md:
```diff
@@ -0,0 +1,137 @@
+OpenStack administrator)
+1. Juju installed (see [Installing Juju][juju-install]), and the OpenStack environment is configured as cloud provider
+
+### Kubernetes
```
I think this section about kubernetes should not be included because it already exists at [1]. I think the tutorial that points to this "how to deploy cdk on openstack" tutorial should actually point to Martin's tutorial (and this file be removed?)

1. https://discourse.charmhub.io/t/how-to-perform-maintenance-on-a-kubernetes-worker/3910

auria commented on juju-tutorials/how-to-deploy-CDK-on-openstack.md:
````diff
@@ -0,0 +1,137 @@
+kubectl create -f <nginx_config>.yaml
+```
+
+### Update QEMU
````
"Update qemu" should not belong in a "how to deploy cdk on openstack" tutorial.

auria reviewed (commented):
Changes look better. I will have a second pass to the whole change tomorrow morning to give the final approval.

Please have a look to my first comment as I think one of the tutorials should not exist (the one about deploying CDK on top of OpenStack). Let me know what you think. I understand it's not exactly the same as Martin's doc, because that one not only deploys CDK but explains how to put a k8s worker in maintenance mode. However, this "Qemu tutorial" needs that maintenance mode as part of the operation, which would match the end goal, wouldn't it?

Thank you!

Commit:
test the entire deployment

- add a note on the need to resolve `ovn-chassis` unit after adding and
  configuring new port
- test and fix all commands

rgildein commented:
I think that the tutorial about about deploying CDK on top of OpenStack, could be removed. However there is a part about creating PVC storage, which shoul be present at least as note. What do you think?

Yes and the part about QEMU needs to be verified (the whole point of this PR) and is still in the WIP.

auria commented:
> I think that the tutorial about about deploying CDK on top of OpenStack, could be removed. However there is a part about creating PVC storage, which shoul be present at least as note. What do you think?

You are right. Martin's tutorial does not show how to create PVC storage.

I think the driving tutorial (the main one you are writing) will reference other tutorials here and there in the "requirements" section. Then, it could show how to create PVC storage in the CDK env. Then it would reference the need to put the k8s worker that runs on the nova-compute unit that we want to upgrade in maintenance mode. And then it will show how to run the qemu package upgrade(s).

Does it sound good to you?

Commit:
remove md file and part of it only as note

Commit:
Fixes on phrasing

rgildein committed:
Merge pull request #1 from auria/review/rgildein/nova-compute-fix-volume-attach

Fixes on phrasing

auria reviewed (approved):
LGTM
